### PR TITLE
tn rename col

### PR DESCRIPTION
**Pull request by aptend**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #14071

## What this PR does / why we need it:

change column name

### CODE REVIEW

**Comment by matrix-meow** [comment]
@aptend Thanks for your contributions!

**Here are review comments for file pkg/pb/api/api.go:**
Title: tn rename col

Body: This pull request aims to change the column name.

Changes in pkg/pb/api/api.go:
- Added a new function `NewRenameColumnReq` that takes in the database ID, table ID, old column name, new column name, and sequence number as parameters. This function returns an `AlterTableReq` struct with the appropriate values set for renaming a column.
- The `NewRemoveColumnReq` function has been modified to include the `seqnum` parameter.

Review:

1. Inconsistent naming convention: The function `NewRenameColumnReq` uses camel case for the parameter names (`oldname` and `newname`), while the rest of the codebase uses snake case. It would be better to maintain consistency and use snake case for all parameter names.

Suggestion: Rename the parameters `oldname` and `newname` to `oldName` and `newName` respectively.

2. Lack of validation: The `NewRenameColumnReq` function does not perform any validation on the input parameters. It would be beneficial to add some validation checks to ensure that the provided database ID, table ID, old column name, and new column name are valid and meet any necessary requirements.

Suggestion: Add validation checks to ensure the validity of the input parameters.

3. Unused function: The `MarshalBinary` function in the `SyncLogTailReq` struct is defined but not used anywhere in the codebase. This can lead to confusion and unnecessary code clutter.

Suggestion: Remove the unused `MarshalBinary` function to improve code clarity.

4. Lack of documentation: The added functions `NewRenameColumnReq` and the modified `NewRemoveColumnReq` lack proper documentation comments. Documentation is essential for understanding the purpose and usage of these functions.

Suggestion: Add documentation comments to the functions to explain their purpose, parameters, and return values.

5. Security concern: The pull request does not mention any security-related changes, but it is important to ensure that the renaming of columns does not introduce any security vulnerabilities. It would be beneficial to conduct a security review of the changes made in this pull request to identify any potential security issues.

Suggestion: Perform a security review of the changes to ensure there are no security vulnerabilities introduced.

Optimization suggestions:

1. Code reuse: The `NewRenameColumnReq` and `NewRemoveColumnReq` functions have similar structures and could potentially share some common code. Consider refactoring the code to reuse common logic and reduce duplication.

2. Error handling: The code does not handle any potential errors that may occur during the execution of the functions. It would be beneficial to add appropriate error handling to handle any potential errors and provide meaningful error messages to the caller.

Overall, the changes made in this pull request address the task of renaming a column. However, there are some areas that need improvement, such as consistent naming conventions, validation checks, documentation, and error handling. Additionally, a security review should be conducted to ensure the changes do not introduce any security vulnerabilities.

**Here are review comments for file pkg/vm/engine/tae/catalog/schema.go:**
Review:

Title: tn rename col

Body: 
- This pull request is a feature request to rename a column.
- It fixes issue #14071.

Changes made in pkg/vm/engine/tae/catalog/schema.go:
- Added a new case for `apipb.AlterKind_RenameColumn` in the `ApplyAlterTable` function.
- Added logic to handle the renaming of a column.
- Checks if the new column name already exists and returns an error if it does.
- Checks if the old column name exists and returns an error if it doesn't.
- Checks if the sequence number of the old column matches the provided sequence number and returns an error if it doesn't.
- Renames the column by updating its name.
- Adds the old column name to the list of dropped attributes.
- Removes the new column name from the list of dropped attributes.
- Logs the renaming of the column.

Potential Problems and Suggestions:
1. Security Issue: The code does not include any validation or sanitization of the column names. This could potentially lead to SQL injection attacks if the column names are not properly validated. It is recommended to validate and sanitize the column names before using them in any SQL queries.

2. Code Optimization: The code currently iterates over all column definitions to find the target column by name. This can be optimized by using a map or a more efficient data structure to store the column definitions, allowing for faster lookup by name.

3. Error Handling: The code currently returns generic internal error messages when encountering issues such as duplicate column names or missing columns. It would be more helpful to provide specific error messages that indicate the exact issue, such as "Column 'newName' already exists" or "Column 'oldName' not found".

4. Logging: The logging statement in the code is not very informative. It would be beneficial to include more details in the log message, such as the table name and any other relevant information.

5. Testing: It would be valuable to include unit tests for the new functionality to ensure that it works as expected and to catch any potential issues or edge cases.

Overall, the changes made in this pull request address the feature request to rename a column. However, there are some potential problems and areas for improvement that should be addressed before merging the code.

**Here are review comments for file pkg/vm/engine/tae/rpc/inspect.go:**
Title: tn rename col

Body: This pull request changes the column name.

Changes in pkg/vm/engine/tae/rpc/inspect.go:
- Import statement for "github.com/matrixorigin/matrixone/pkg/vm/engine/tae/iface/handle" has been added.
- A new struct `RenameColArg` has been added, which represents the arguments for renaming a column.
- Methods `FromCommand`, `String`, and `Run` have been added to the `RenameColArg` struct.
- The `initCommand` function has been modified to add a new command `rename_col` for renaming a column.

Overall, the changes seem to be focused on adding functionality to rename a column in the inspect command.

Potential Problems:
1. Security: The `RenameColArg` struct does not perform any input validation on the `oldName`, `newName`, and `seq` fields. This could potentially lead to SQL injection or other security vulnerabilities. It is important to validate and sanitize user input before using it in SQL queries.

Suggestions:
1. Security: Implement input validation and sanitization for the `oldName`, `newName`, and `seq` fields in the `RenameColArg` struct. This can be done by checking for valid column names and ensuring that the `seq` value is within the valid range.
2. Code Organization: Consider separating the `RenameColArg` struct and its methods into a separate file for better code organization and maintainability.
3. Documentation: Add comments or documentation to explain the purpose and usage of the `RenameColArg` struct and its methods.
4. Error Handling: Add proper error handling in the `Run` method of the `RenameColArg` struct. Currently, if an error occurs during the renaming process, the transaction is rolled back, but the error is not returned or logged.

Optimizations:
1. Consider using a more descriptive name for the `RenameColArg` struct to improve code readability.
2. Consider using a more descriptive name for the `RunFactory` function to clarify its purpose.
3. Consider using constants or variables instead of hardcoding values like `0` and `"*"` to improve code maintainability and readability.

**Here are review comments for file pkg/vm/engine/tae/txn/txnimpl/table.go:**
Title: tn rename col

Body: This pull request is a feature request that addresses issue #14071. It changes the column name in the file `pkg/vm/engine/tae/txn/txnimpl/table.go`.

Changes in `pkg/vm/engine/tae/txn/txnimpl/table.go`:
- Added `apipb.AlterKind_RenameColumn` to the list of valid alter kinds in the `AlterTable` function.

Review:

The changes made in this pull request seem to be straightforward and address the specific issue mentioned. However, there are a few areas that could be improved:

1. Lack of detailed explanation: The pull request title and body do not provide sufficient information about why the column name is being changed. It would be helpful to include a brief explanation of the reason behind this change.

2. Inconsistent naming: The pull request title mentions "tn rename col," which is not clear and may be confusing. It would be better to use a more descriptive title that clearly conveys the purpose of the change.

3. Missing error handling: The `AlterTable` function does not handle the case when an unsupported alter kind is provided. It would be beneficial to add appropriate error handling or validation to handle this scenario.

4. Security considerations: Based on the provided information, there don't appear to be any immediate security issues. However, it is always important to consider potential security implications when making changes to code. It would be advisable to review the code changes thoroughly to ensure that they do not introduce any security vulnerabilities.

To optimize the changes made in this pull request, consider the following suggestions:

1. Provide a more descriptive title: Instead of "tn rename col," consider using a title that clearly describes the purpose of the change, such as "Rename column in txn table implementation."

2. Include a detailed explanation: Add a more detailed explanation in the pull request body, describing why the column name is being changed and any potential impact or benefits of the change.

3. Handle unsupported alter kinds: Update the `AlterTable` function to handle unsupported alter kinds gracefully. This could involve returning an appropriate error message or rejecting the request with a clear error response.

4. Review for security implications: Conduct a thorough review of the code changes to ensure that they do not introduce any security vulnerabilities. Consider potential edge cases and validate inputs to prevent any security risks.

Overall, the changes made in this pull request seem to be reasonable, but there is room for improvement in terms of clarity, error handling, and security considerations.

**Here are review comments for file proto/api.proto:**
Review:

Title: tn rename col

Body: This pull request aims to rename a column.

Changes made in the `api.proto` file:
- Added a new enum value `RenameColumn` to the `AlterKind` enum.
- Added a new message `AlterTableRenameCol` to handle column renaming. It includes fields for the old column name, new column name, and a sequence number for double-checking.

Overall, the changes made in this pull request seem to be straightforward and address the specific task of renaming a column. However, there are a few areas that could be improved:

1. **Missing context in the pull request description**: The pull request description lacks context about why the column needs to be renamed. Providing a brief explanation of the reason behind the renaming would help reviewers understand the motivation behind the change.

2. **Lack of validation for column renaming**: The current implementation of column renaming does not include any validation. It would be beneficial to add validation checks to ensure that the old column name exists and that the new column name is unique within the table schema. This will prevent potential issues and ensure data integrity.

3. **Security concern: Sequence number usage**: The `sequence_num` field in the `AlterTableRenameCol` message is mentioned as being used for double-checking. However, it is not clear how this field is used or what purpose it serves. If this field is intended for security purposes, it is important to provide more details about its usage and potential security implications.

To optimize the changes made in this pull request, consider the following suggestions:

1. **Include unit tests**: Adding unit tests for the column renaming functionality will help ensure that it works as expected and prevent regressions in the future.

2. **Provide examples or usage documentation**: Including examples or usage documentation in the pull request description or relevant documentation files will help other developers understand how to use the column renaming feature correctly.

3. **Consider code refactoring**: While the changes made in this pull request are minimal, it may be worth considering if there are any opportunities for code refactoring or simplification in the affected codebase. This could help improve code readability and maintainability.

Overall, this pull request addresses the task of renaming a column, but there are areas that need improvement, such as providing more context, adding validation checks, and clarifying the usage of the `sequence_num` field. Additionally, including unit tests and documentation will help ensure the feature is robust and easy to use.

**Review by XuPeng-SH — APPROVED** [review]

**Review by zhangxu19830126 — APPROVED** [review]

**Pull request merged**
